File: scripts/test_capy_websocket.py
#!/usr/bin/env python3
"""Exercise real nginx/FastCGI text, binary, and close paths for a Capy WS handler."""

import base64
import os
import socket
import struct

HOST = "bearer.openfu.com"
PATH = "/tests/capy-websocket.capy"


def receive_exact(sock: socket.socket, size: int) -> bytes:
    data = bytearray()
    while len(data) < size:
        chunk = sock.recv(size - len(data))
        if not chunk:
            raise RuntimeError("WebSocket closed before the expected frame")
        data.extend(chunk)
    return bytes(data)


def receive_frame(sock: socket.socket) -> tuple[int, bytes]:
    first, second = receive_exact(sock, 2)
    length = second & 0x7F
    if length == 126:
        length = struct.unpack("!H", receive_exact(sock, 2))[0]
    elif length == 127:
        length = struct.unpack("!Q", receive_exact(sock, 8))[0]
    return first & 0x0F, receive_exact(sock, length)


def exchange(opcode: int, payload: bytes, expected: list[tuple[int, bytes]]) -> None:
    with socket.create_connection(("127.0.0.1", 80), timeout=10) as connection:
        key = base64.b64encode(os.urandom(16)).decode()
        request = (
            f"GET {PATH} HTTP/1.1\r\nHost: {HOST}\r\nUpgrade: websocket\r\n"
            f"Connection: Upgrade\r\nSec-WebSocket-Key: {key}\r\nSec-WebSocket-Version: 13\r\n\r\n"
        )
        connection.sendall(request.encode())
        response = bytearray()
        while b"\r\n\r\n" not in response:
            response.extend(connection.recv(4096))
        if not response.startswith(b"HTTP/1.1 101 "):
            raise SystemExit(f"Capy WebSocket upgrade failed: {response[:200]!r}")

        mask = b"\x12\x34\x56\x78"
        masked = bytes(value ^ mask[index % 4] for index, value in enumerate(payload))
        connection.sendall(bytes([0x80 | opcode, 0x80 | len(payload)]) + mask + masked)
        frames = [receive_frame(connection) for _ in expected]
        if frames != expected:
            raise SystemExit(f"Capy WebSocket frame mismatch: {frames!r}, expected {expected!r}")


exchange(1, b"hello", [(1, b"echo:hello"), (1, b"direct:hello"), (8, b"\x03\xe8")])
exchange(2, b"\x00\xffcapy", [(2, b"\x00\xffcapy"), (8, b"\x03\xe8")])
print("Capy WebSocket text/binary inspect/send/close path passed")

File: scripts/test_capy_websocket.cpp
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/socket.h>
#include <unistd.h>
#include <array>
#include <cstring>
#include <stdexcept>
#include <string>
#include <vector>

static void send_all(int fd, const std::string& data) { for (size_t at=0;at<data.size();) { auto n=send(fd,data.data()+at,data.size()-at,0); if(n<=0) throw std::runtime_error("WebSocket send failed"); at+=n; } }
static std::string receive_exact(int fd, size_t size) { std::string out(size, '\0'); for(size_t at=0;at<size;) { auto n=recv(fd,out.data()+at,size-at,0); if(n<=0) throw std::runtime_error("WebSocket closed before expected frame"); at+=n; } return out; }
static std::pair<unsigned,std::string> frame(int fd) { auto h=receive_exact(fd,2); size_t n=unsigned(h[1])&127; if(n==126){auto x=receive_exact(fd,2);n=(unsigned((unsigned char)x[0])<<8)|(unsigned char)x[1];} else if(n==127) throw std::runtime_error("unexpected oversized frame"); return {unsigned(h[0])&15,receive_exact(fd,n)}; }
static void exchange(unsigned opcode, const std::string& payload, const std::vector<std::pair<unsigned,std::string>>& expected) { int fd=socket(AF_INET,SOCK_STREAM,0); sockaddr_in addr{};addr.sin_family=AF_INET;addr.sin_port=htons(80);inet_pton(AF_INET,"127.0.0.1",&addr.sin_addr);if(connect(fd,(sockaddr*)&addr,sizeof(addr)))throw std::runtime_error("WebSocket connect failed"); send_all(fd,"GET /tests/capy-websocket.capy HTTP/1.1\r\nHost: bearer.openfu.com\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Key: MTIzNDU2Nzg5MDEyMzQ1Ng==\r\nSec-WebSocket-Version: 13\r\n\r\n");std::string response;while(response.find("\r\n\r\n")==std::string::npos)response+=receive_exact(fd,1);if(!response.starts_with("HTTP/1.1 101 "))throw std::runtime_error("Capy WebSocket upgrade failed");std::string f;f.push_back(char(128|opcode));f.push_back(char(128|payload.size()));const std::array<unsigned char,4> mask={0x12,0x34,0x56,0x78};f.append((const char*)mask.data(),4);for(size_t i=0;i<payload.size();++i)f.push_back(payload[i]^mask[i%4]);send_all(fd,f);for(auto& want:expected)if(frame(fd)!=want)throw std::runtime_error("Capy WebSocket frame mismatch");close(fd); }
int main(){try{exchange(1,"hello",{{1,"echo:hello"},{1,"direct:hello"},{8,std::string("\x03\xe8",2)}});exchange(1,"flush",{{1,"flush-rejected"},{8,std::string("\x03\xe8",2)}});exchange(2,std::string("\0\xff" "capy",6),{{2,std::string("\0\xff" "capy",6)},{8,std::string("\x03\xe8",2)}});}catch(const std::exception& e){write(2,e.what(),std::strlen(e.what()));write(2,"\n",1);return 1;}return 0;}
